src/KubernetesClient/Informers/SharedInformer.cs

```diff
@@ -0,0 +1,184 @@
+        {
+            _cache = cache;
+            _masterScheduler = scheduler ?? new EventLoopScheduler();
+            _logger = logger;
+            _keySelector = keySelector;
+            _masterObservable = masterInformer
+                .GetResource(ResourceStreamType.ListWatch)
+                .ObserveOn(_masterScheduler)
+                .Do(x => _logger.LogTrace($"Received message from upstream {x}"))
+                .SynchronizeCache(_cache, _keySelector)
+                .Do(msg =>
+                {
+                    // cache is synchronized as soon as we get at least one message past this point
+                    _logger.LogTrace($"Cache v{cache.Version} synchronized: {msg} ");
+                    _cacheSynchronized.TrySetResult(true);
+                    _logger.LogTrace("_cacheSynchronized.TrySetResult(true)");
+                })
+                .Do(_ => YieldToWaitingSubscribers())
+                .ObserveOn(Scheduler.Immediate) // immediate ensures that all caches operations are done atomically
+                .ObserveOn(_masterScheduler)
+                .Catch<CacheSynchronized<ResourceEvent<TResource>>, Exception>(e =>
+                {
+                    _cacheSynchronized.TrySetException(e);
+                    // _cacheSynchronized.OnError(e);
+                    return Observable.Throw<CacheSynchronized<ResourceEvent<TResource>>>(e);
+                })
+                .Finally(() => _cacheSynchronized.TrySetResult(false))
+                // .SubscribeOn(_masterScheduler)
+                .Publish();
+        }
+
+
+        public IObservable<ResourceEvent<TResource>> GetResource(ResourceStreamType type)
+        {
+            var childScheduler = new EventLoopScheduler(); // dedicated thread for the child on which all messages are syncrhonized
+            return Observable.Defer(async () =>
+                {
+                    AddSubscriber();
+                    _logger.LogTrace("Subscriber awaiting cache synchronization before attaching");
+
+                    var isCacheSynchronized = await _cacheSynchronized.Task;
+                    if (!isCacheSynchronized) // really this only happens if the reset is the master completes before first reset, in which case the downstream subscriber gets nothing
+                        return Observable.Empty<ResourceEvent<TResource>>();
+                    // we use lock to pause any processing of the broadcaster while we're attaching to the stream so proper alignment can be made
+
+                    _logger.LogTrace("Subscriber attaching to broadcaster");
+                    return Observable.Create<ResourceEvent<TResource>>(observer =>
+                        {
+                            var broadcasterAttachment = Disposable.Empty;
+                            var cacheSnapshot = _cache.Snapshot();
```

Comment:
Is it necessary to snapshot the cache? It seems we only need to capture the version of the cache.